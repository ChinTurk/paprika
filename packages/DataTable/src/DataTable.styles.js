import { css } from "styled-components";

export const dummyTableStyles = css`
  background: white;
`;
