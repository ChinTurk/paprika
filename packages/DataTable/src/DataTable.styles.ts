import styled from "styled-components";

export const Wrapper = styled.div`
  width: 100%;
  /* overflow: scroll; */

  .table {
    border: 1px solid #ddd;
    width: 100%;
    /* overflow: scroll; */

    .tr {
      :last-child {
        .td {
          border-bottom: 0;
        }
      }
    }

    .th,
    .td {
      padding: 5px;
      border-bottom: 1px solid #ddd;
      border-right: 1px solid #ddd;
      background-color: #fff;
      overflow: hidden;

      :last-child {
        border-right: 0;
      }

      .resizer {
        display: inline-block;
        width: 5px;
        height: 100%;
        position: absolute;
        right: 0;
        top: 0;
        transform: translateX(50%);
        z-index: 1;

        &.isResizing {
          background: red;
        }
      }
    }

    &.sticky {
      /* overflow: scroll; */
      .header,
      .footer {
        position: sticky;
        z-index: 1;
        width: fit-content;
      }

      .header {
        z-index: 4;
        top: 0;
        box-shadow: 0px 3px 3px #ccc;
      }

      .footer {
        z-index: 4;
        bottom: 0;
        box-shadow: 0px -3px 3px #ccc;
      }

      .body {
        position: relative;
      }

      [data-sticky-td] {
        position: sticky;
      }

      [data-sticky-last-left-td] {
        box-shadow: 2px 0px 3px #ccc;
      }

      [data-sticky-first-right-td] {
        box-shadow: -2px 0px 3px #ccc;
      }
    }
  }
`;
