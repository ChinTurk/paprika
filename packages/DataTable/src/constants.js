export const actions = {
  RESET_DATA: "RESET_DATA",
  SORT: "SORT",
};

export const sortDirections = {
  ASCEND: "ASCEND",
  DESCEND: "DESCEND",
};

export const columnTypes = {
  TEXT: "TEXT",
  NUMBER: "NUMBER",
  DATE: "DATE",
};
