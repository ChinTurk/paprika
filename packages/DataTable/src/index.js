export { useDataTableState, useDispatch } from "./context";
export { default as RowHeight } from "./components/RowHeight";
export { default as Sort } from "./components/Navigation/components/Sort";
export { default } from "./DataTable";
