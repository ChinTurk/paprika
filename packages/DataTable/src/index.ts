export { default as Table } from "./DataTable";
export * from "./DataTable";

export { InfiniteLoader } from "./components/InfiniteLoader";
