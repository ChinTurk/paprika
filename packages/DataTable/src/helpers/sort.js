import isString from "lodash.isstring";
import { sortDirections } from "../constants";

function compareString(a, b) {
  return a.localeCompare(b);
}

function compareNumber(a, b) {
  return a - b;
}

export default function sort({ data, keygen, columnId, direction, hasBackendSort = false }) {
  if (hasBackendSort) return null;

  switch (direction) {
    case sortDirections.ASCEND:
      return [...data]
        .sort((rowA, rowB) =>
          isString(rowA[columnId])
            ? compareString(rowA[columnId], rowB[columnId])
            : compareNumber(rowA[columnId], rowB[columnId])
        )
        .map(item => item[keygen]);
    case sortDirections.DESCEND:
      return [...data]
        .sort((rowA, rowB) =>
          isString(rowA[columnId])
            ? -compareString(rowA[columnId], rowB[columnId])
            : -compareNumber(rowA[columnId], rowB[columnId])
        )
        .map(item => item[keygen]);
    default:
      return null;
  }
}
