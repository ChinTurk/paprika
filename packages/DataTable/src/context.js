import React from "react";
import PropTypes from "prop-types";
import isMatchWith from "lodash.ismatchwith";
import tableReducer from "./reducers/table";
import useAsyncReducer from "./hooks/useAsyncReducer";
import ruleTesters from "./components/Navigation/components/Filters/ruleTesters";
import sort from "./helpers/sort";
import { columnTypes } from "./constants";

const TableStateContext = React.createContext();
const TableDispatchContext = React.createContext();

function TableProvider(props) {
  const { data, keygen, reducers, columns } = props;
  const initialState = {
    data: data || [],
    dataForRendering: data || [],
    keygen,
    sortColumn: null,
    sortDirection: null,
    columnsOrder: columns.map(column => column.id),
    columns: columns.reduce((columnsObject, column) => ({ ...columnsObject, [column.id]: column }), {}),
    filters: [],
  };

  const isFirstRender = React.useRef(true);
  const memorizedReducer = React.useCallback(
    (state, action) => {
      const changes = tableReducer(state, action);
      return new Promise(resolve => {
        resolve(
          reducers.reduce(async (previousPromise, reducer) => {
            const prevChanges = await previousPromise;
            const result = reducer(state, { ...action, changes: prevChanges });
            if (typeof result.then === "function") {
              const finalResult = await result;
              return finalResult;
            }
            return result;
          }, changes)
        );
      });
    },
    [reducers]
  );
  const [state, dispatch] = useAsyncReducer(memorizedReducer, initialState);

  React.useEffect(() => {
    if (isFirstRender.current) {
      isFirstRender.current = false;
      return;
    }

    dispatch({ type: "RESET_STATE", payload: { data } });
    // Watching data
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [data, isFirstRender]);

  React.useEffect(() => {
    const newColumns = columns.reduce((columnsObject, column) => ({ ...columnsObject, [column.id]: column }), {});
    const isVisibilityChanged = !isMatchWith(newColumns, state.columns, (newColumn, column) => {
      return newColumn.id === column.id && newColumn.isHidden === column.isHidden;
    });

    if (isVisibilityChanged) {
      dispatch({
        type: "RESET_STATE",
        payload: {
          columns: newColumns,
        },
      });
    }
    // Watching isHidden
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [columns]);

  return (
    <TableStateContext.Provider value={state}>
      <TableDispatchContext.Provider value={dispatch}>{props.children}</TableDispatchContext.Provider>
    </TableStateContext.Provider>
  );
}

function useDataTableState() {
  return React.useContext(TableStateContext);
}

function useDispatch() {
  return React.useContext(TableDispatchContext);
}

function useSortedAndFilteredData() {
  const { data, columns, filters, sortColumn, sortDirection, keygen } = useDataTableState();
  let sortedData = [];
  let filteredData = [];

  function calculateResult() {
    if (sortColumn && sortDirection) {
      sortedData = sort({
        data,
        columnId: sortColumn,
        direction: sortDirection,
        columnType:
          columns[sortColumn].type || (typeof data[0][sortColumn] === "number" ? columnTypes.NUMBER : columnTypes.TEXT),
        momentParsingFormat: columns[sortColumn].momentParsingFormat,
      });
    }

    if (filters.length > 0) {
      filteredData = data.filter(row =>
        filters.every(filter => ruleTesters[filter.rule](row[filter.columnId], filter.value))
      );
    }

    if (sortedData.length === 0 && filteredData.length === 0) return data;

    if (sortedData.length > 0 && filteredData.length > 0)
      return sortedData.filter(row => filteredData.find(filteredRow => row[keygen] === filteredRow[keygen]));

    return sortedData.length > 0 ? sortedData : filteredData;
  }

  const result = React.useMemo(calculateResult, [data, columns, filters, sortColumn, sortDirection, keygen]);

  return result;
}

TableProvider.propTypes = {
  children: PropTypes.node.isRequired,
  data: PropTypes.arrayOf(PropTypes.objectOf(PropTypes.any)),
  keygen: PropTypes.string.isRequired,
  reducers: PropTypes.arrayOf(PropTypes.func).isRequired,
  columns: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.oneOfType([PropTypes.number, PropTypes.string]).isRequired,
    })
  ),
};

TableProvider.defaultProps = {
  data: [],
  columns: [],
};

export { TableProvider, useDataTableState, useDispatch, useSortedAndFilteredData };
