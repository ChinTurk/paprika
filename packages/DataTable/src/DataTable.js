/* eslint-disable react/prop-types */
import React from "react";
import PropTypes from "prop-types";
import nanoid from "nanoid";
import ColumnDefinition from "./components/ColumnDefinition";
import Navigation from "./components/Navigation";
import VirtualizedTable from "./components/VirtualizedTable";
import LoadMoreButton from "./components/LoadMoreButton";
import { extractChildren } from "./helpers";
import { sortDirections, columnTypes } from "./constants";
import { TableProvider } from "./context";

const propTypes = {
  data: PropTypes.arrayOf(PropTypes.shape({})),
  height: PropTypes.number,
  width: PropTypes.number,
  rowHeight: PropTypes.number,
  reducers: PropTypes.arrayOf(PropTypes.func),
  isLoading: PropTypes.bool,
  onExpandedRow: PropTypes.func,
  onKeyDownArrow: PropTypes.func,
  onClickCell: PropTypes.func,
};

const defaultProps = {
  data: [],
  height: 600,
  width: 640,
  rowHeight: 32,
  reducers: [],
  isLoading: false,
  onExpandedRow: () => {},
  onKeyDownArrow: () => {},
  onClickCell: () => {},
};

export default function DataTable(props) {
  const dataTableID = React.useRef(`DT${nanoid()}DTCELL`).current;
  const {
    children: childrenProps,
    data,
    height,
    isLoading,
    keygen,
    onKeyDownArrow,
    onClickCell,
    onExpandedRow,
    reducers,
    rowHeight,
    width,
  } = props;
  const {
    "DataTable.ColumnDefinition": ColumnsDefinition,
    "DataTable.Navigation": Navigation,
    "DataTable.LoadMoreButton": LoadMoreButton,
  } = extractChildren(childrenProps, [
    "DataTable.ColumnDefinition",
    "DataTable.Navigation",
    "DataTable.LoadMoreButton",
  ]);

  const columns = ColumnsDefinition.map(Column => Column.props);

  let navigationReducers = [];
  let isControlled = false;
  if (Navigation && Navigation.props) {
    navigationReducers = React.Children.map(Navigation.props.children, child => child.type.reducer).filter(
      chunk => chunk
    );
    React.Children.forEach(Navigation.props.children, child => {
      if (child.props.onFilter || child.props.onSort) isControlled = true;
    });
  }

  return (
    <TableProvider
      isControlled={isControlled}
      data={data}
      keygen={keygen}
      reducers={navigationReducers.concat(reducers)}
      columns={columns}
    >
      <div>{isLoading ? "Loading..." : null}</div>
      {Navigation}
      <VirtualizedTable
        dataTableID={dataTableID}
        onExpandedRow={onExpandedRow}
        columns={columns}
        height={height}
        rowHeight={rowHeight}
        width={width}
        LoadMoreButton={LoadMoreButton}
        onKeyDownArrow={onKeyDownArrow}
        onClickCell={onClickCell}
      />
    </TableProvider>
  );
}

DataTable.prpoTypes = propTypes;
DataTable.defaultProps = defaultProps;
DataTable.ColumnDefinition = ColumnDefinition;
DataTable.SortDirections = { ...sortDirections };
DataTable.Navigation = Navigation;
DataTable.ColumnTypes = columnTypes;
DataTable.LoadMoreButton = LoadMoreButton;
