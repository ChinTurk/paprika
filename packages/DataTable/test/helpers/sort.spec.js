import sort from "../../src/helpers/sort";
import { sortDirections } from "../../src/constants";

const originalData = [
  {
    id: 1,
    salary: 3000,
    name: "Abc",
  },
  {
    id: 2,
    salary: 1000,
    name: "Cba",
  },
  {
    id: 3,
    salary: 2000,
    name: "Acb",
  },
];

describe("helpers/sort", () => {
  it("sorts an array of items with number ascend", () => {
    expect(
      sort({
        data: originalData,
        columnId: "salary",
        direction: sortDirections.ASCEND,
      })
    ).toMatchSnapshot();
  });

  it("sorts an array of items with number descend", () => {
    expect(
      sort({
        data: originalData,
        columnId: "salary",
        direction: sortDirections.DESCEND,
      })
    ).toMatchSnapshot();
  });

  it("sorts an array of items with string ascend", () => {
    expect(
      sort({
        data: originalData,
        columnId: "name",
        direction: sortDirections.ASCEND,
      })
    ).toMatchSnapshot();
  });

  it("sorts an array of items with string descend", () => {
    expect(
      sort({
        data: originalData,
        columnId: "name",
        direction: sortDirections.DESCEND,
      })
    ).toMatchSnapshot();
  });
});
