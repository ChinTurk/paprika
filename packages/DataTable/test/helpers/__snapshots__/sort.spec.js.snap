// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`helpers/sort sorts an array of items with number ascend 1`] = `
Array [
  Object {
    "id": 2,
    "name": "Cba",
    "salary": 1000,
  },
  Object {
    "id": 3,
    "name": "Acb",
    "salary": 2000,
  },
  Object {
    "id": 1,
    "name": "Abc",
    "salary": 3000,
  },
]
`;

exports[`helpers/sort sorts an array of items with number descend 1`] = `
Array [
  Object {
    "id": 1,
    "name": "Abc",
    "salary": 3000,
  },
  Object {
    "id": 3,
    "name": "Acb",
    "salary": 2000,
  },
  Object {
    "id": 2,
    "name": "Cba",
    "salary": 1000,
  },
]
`;

exports[`helpers/sort sorts an array of items with string ascend 1`] = `
Array [
  Object {
    "id": 1,
    "name": "Abc",
    "salary": 3000,
  },
  Object {
    "id": 3,
    "name": "Acb",
    "salary": 2000,
  },
  Object {
    "id": 2,
    "name": "Cba",
    "salary": 1000,
  },
]
`;

exports[`helpers/sort sorts an array of items with string descend 1`] = `
Array [
  Object {
    "id": 2,
    "name": "Cba",
    "salary": 1000,
  },
  Object {
    "id": 3,
    "name": "Acb",
    "salary": 2000,
  },
  Object {
    "id": 1,
    "name": "Abc",
    "salary": 3000,
  },
]
`;
